common/analysis/lint_rule_status.h

```diff
@@ -28,27 +28,101 @@
 #include "common/text/symbol.h"
 #include "common/text/syntax_tree_context.h"
 #include "common/text/token_info.h"
+#include "common/util/logging.h"
 
 namespace verible {
 
+// Represents a single replace operation on a text fragment.
+//
+// Either fragment or replacement can be strings with zero width, providing a
+// way for, respectively, inserting and removing text.
+//
+// ReplacementEdit differs from editscript's Edit in that it stores a
+// replacement string, so it doesn't need the "after" text to be useful.
+struct ReplacementEdit {
+  ReplacementEdit(absl::string_view fragment, const std::string& replacement)
+      : fragment(fragment), replacement(replacement) {}
+
+  ReplacementEdit(const TokenInfo& token, const std::string& replacement)
+      : fragment(token.text()), replacement(replacement) {}
+
+  bool operator<(const ReplacementEdit& other) const {
+    // Check that the fragment is located before the other's fragment. When they
+    // overlap, `this<other` and `other<this` return false, which makes them
+    // equivalent in std::set.
+    return (fragment.data() + fragment.size()) <= other.fragment.data();
+  }
+
+  absl::string_view fragment;
+  std::string replacement;
+};
+
+// Collection of ReplacementEdits performing single violation fix.
+class AutoFix {
+ public:
+  AutoFix() : description(), edits() {}
+  AutoFix(const AutoFix& other)
+      : description(other.description), edits(other.edits) {}
+  AutoFix(const AutoFix&& other)
+      : description(std::move(other.description)),
+        edits(std::move(other.edits)) {}
+
+  AutoFix(const std::string& description,
+          std::initializer_list<ReplacementEdit> edits)
+      : description(description), edits(edits) {
+    CHECK_EQ(this->edits.size(), edits.size()) << "Edits must not overlap.";
+  }
+
+  AutoFix(std::initializer_list<ReplacementEdit> edits)
+      : AutoFix("Fix", edits) {}
+
+  AutoFix(const std::string& description, ReplacementEdit edit)
+      : AutoFix(description, {edit}) {}
+
+  AutoFix(ReplacementEdit edit) : AutoFix({edit}) {}
+
+  // Applies the fix on a `base` and returns modified text.
+  std::string Apply(const absl::string_view base) const;
```

Comment:
It should not fail as long as `base` is the same string as the one used for creating edits.
I've added `CHECK_*` for making sure this is the case.

Since edit conflict detection takes place in `AddEdit` and a constructor, I've also added additional test cases for testing that these checks work as intended.